sirepo/template/radia.py

```diff
@@ -1092,51 +1126,179 @@ def _save_kick_map_sdds(name, x_vals, y_vals, h_vals, v_vals, path):
-    c = [numpy.sum(w * ctr), numpy.sum(h * ctr)]
-    s = [numpy.sum(w * sz), numpy.sum(h * sz)]
+    o.points = _update_extrusion_points(
+        [
+            [ax1, ay1], [ax2, ay1], [ax2, ay2],
+            [sx2, ay2], [sx2, sy2], [ax2, sy2], [ax2, sy1], [sx1, sy1],
+            [ax1, ay1]
+        ],
+        c,
+        [int(o.stemPosition), int(o.armPosition)],
+        w,
+        b
+    )
+    return o
 
-    # start with arm top, stem left - then reflect across centroid axes as needed
-    ax1 = c[0] - s[0] / 2
-    ax2 = ax1 + s[0]
-    ay1 = c[1] + s[1] / 2
-    ay2 = ay1 - ell.armHeight
 
-    sx1 = c[0] - s[0] / 2
-    sx2 = sx1 + ell.stemWidth
-    sy = c[1] - s[1] / 2
-
-    k = [int(ell.stemPosition), int(ell.armPosition)]
-    ell.points = [
-        [ax1, ay1], [ax2, ay1], [ax2, ay2],
-        [sx2, ay2], [sx2, sy], [sx1, sy],
-        [ax1, ay1]
+def _update_cuboid(o, _, __, **kwargs):
+    return _update_geom_obj(o, delim_fields=PKDict(segments=[1, 1, 1]), **kwargs)
+
+
+def _update_ell(o, beam_axis, height_axis, **kwargs):
+    return _update_ell_points(_update_geom_obj(o, **kwargs), beam_axis, height_axis)
+
+
+def _update_ell_points(o, beam_axis, height_axis):
+    g = _update_stemmed_points(o, beam_axis, height_axis)
+    w, h, b, c, s = g.geom
+    ax1, ax2, ay1, ay2, sx1, sx2, sy1 = g.points
+
+    o.points = _update_extrusion_points(
+        [
+            [ax1, ay1], [ax2, ay1], [ax2, ay2],
+            [sx2, ay2], [sx2, sy1], [sx1, sy1],
+            [ax1, ay1]
+        ],
+        c,
+        [int(o.stemPosition), int(o.armPosition)],
+        w,
+        b
+    )
+    return o
+
+
+def _update_jay_points(o, beam_axis, height_axis):
+    g = _update_stemmed_points(o, beam_axis, height_axis)
+    w, h, b, c, s = g.geom
+    ax1, ax2, ay1, ay2, sx1, sx2, sy1 = g.points
+    jx1 = c[0] + s[0] / 2 - o.hookWidth
+    jy1 = ay2 - o.hookHeight
+
+    o.points = _update_extrusion_points(
+        [
+            [ax1, ay1], [ax2, ay1], [ax2, jy1], [jx1, jy1], [jx1, ay2],
+            [sx2, ay2], [sx2, sy1], [sx1, sy1],
+            [ax1, ay1]
+        ],
+        c,
+        [int(o.stemPosition), int(o.armPosition)],
+        w,
+        b
+    )
+    return o
+
+
+def _update_extrusion_points(points, ctr, stem_indices, width_dir, length_dir):
+    pts = [
+        [2 * ctr[i] * stem_indices[i] + (-1)**stem_indices[i] * v for (i, v) in enumerate(p)] \
+        for p in points
     ]
-    ell.points = [[2 * c[i] * k[i] + -1**k[i] * v for (i, v) in enumerate(p)] for p in ell.points]
 
     # Radia's extrusion expects points in permutation order based on the extrusion
     # axis (x -> [y, z], y -> [z, x], z -> [x, y])
-    if w.tolist().index(1) != (b.tolist().index(1) + 1) % 3:
-        for p in ell.points:
+    if width_dir.tolist().index(1) != (length_dir.tolist().index(1) + 1) % 3:
+        for p in pts:
             p.reverse()
 
+    return pts
+
+
+def _update_geom_from_dipole(geom, dipole, beam_axis, height_axis):
```

Comment:
This function is quite large and doing a lot of work. I think it could benefit from calling more functions to make it clearer to the untrained user (me) what the math operations are doing.

Reply:
agree, see recent comment